finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+        assert!(addresses.is_empty());
+        assert!(maybe_command.is_none());
+    }
+
+    #[tokio::test]
+    async fn does_not_rebroadcast_quickly() {
+        let (mut discovery, mut handlers) = build().await;
+        let authentication = handlers[1].authentication();
+        let handler = &mut handlers[0];
+        discovery.handle_message(
+            DiscoveryMessage::AuthenticationBroadcast(authentication.clone()),
+            handler,
+        );
+        let (addresses, maybe_command) = discovery.handle_message(
+            DiscoveryMessage::AuthenticationBroadcast(authentication.clone()),
+            handler,
+        );
+        assert_eq!(addresses.len(), authentication.0.address().len());
+        assert_eq!(
+            addresses[0].encode(),
+            authentication.0.address()[0].encode()
+        );
+        assert!(maybe_command.is_none());
+    }
+
+    #[tokio::test]
+    async fn rebroadcasts_after_cooldown() {
+        let (mut discovery, mut handlers) = build().await;
+        let authentication = handlers[1].authentication();
+        let handler = &mut handlers[0];
+        discovery.handle_message(
+            DiscoveryMessage::AuthenticationBroadcast(authentication.clone()),
+            handler,
+        );
+        sleep(time::Duration::from_millis(MS_COOLDOWN + 5));
+        let (addresses, maybe_command) = discovery.handle_message(
+            DiscoveryMessage::AuthenticationBroadcast(authentication.clone()),
+            handler,
+        );
+        assert_eq!(addresses.len(), authentication.0.address().len());
+        assert_eq!(
+            addresses[0].encode(),
+            authentication.0.address()[0].encode()
+        );
+        match maybe_command {
+            Some((
+                DiscoveryMessage::AuthenticationBroadcast(rebroadcast_authentication),
+                DataCommand::Broadcast,
+            )) => assert_eq!(authentication.encode(), rebroadcast_authentication.encode()),
+            _ => panic!("Expected a rebroadcast, got {:?}", maybe_command),
+        }
+    }
+
+    #[tokio::test]
+    async fn responds_to_correct_request_when_can() {
+        let (mut discovery, mut handlers) = build().await;
+        let requested_authentication = handlers[1].authentication();
+        let requested_node_id = requested_authentication.0.creator();
+        let requester_authentication = handlers[2].authentication();
+        let handler = &mut handlers[0];
+        assert!(handler.handle_authentication(requested_authentication.clone()));
+        let (addresses, maybe_command) = discovery.handle_message(
+            DiscoveryMessage::Request(vec![requested_node_id], requester_authentication.clone()),
+            handler,
+        );
+        assert_eq!(addresses.len(), requester_authentication.0.address().len());
+        assert_eq!(
+            addresses[0].encode(),
+            requester_authentication.0.address()[0].encode()
+        );
+        match maybe_command {
```

Comment:
there is also identical piece of code below, and a few very similar; I know that these are tests etc., but just consider whether there is an easy way of simplifying things here; if not, then there is no problem ofc

Reply:
These have small annoying differences between them, I think implementing a general abstraction over them would be slightly more work than it's worth.